src/gt4py/_core/definitions.py

```diff
@@ -0,0 +1,317 @@
+
+    Additionally, instances of this class can also be used as valid NumPy
+    `dtype`s definitions due to the `.dtype` attribute.
+    """
+
+    scalar_type: type[ScalarT]
+    subshape: tuple[int, ...] = dataclasses.field(default=())
+
+    @functools.cached_property
+    def kind(self) -> DTypeKind:
+        return dtype_kind(self.scalar_type)
+
+    @property
+    def subndim(self) -> int:
+        return len(self.subshape)
+
+    @property
+    def dtype(self) -> np.dtype:
+        """NumPy dtype corresponding to this DType."""  # noqa: D403
+        return np.dtype(self.scalar_type)
+
+
+@dataclasses.dataclass(frozen=True)
+class IntegerDType(DType[IntegerT]):
+    kind: Final[Literal[DTypeKind.INT]] = dataclasses.field(default=DTypeKind.INT, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class UnsignedIntDType(DType[UnsignedIntT]):
+    pass
+
+
+@dataclasses.dataclass(frozen=True)
+class UInt8DType(UnsignedIntDType[uint8]):
+    scalar_type: Final[type[uint8]] = dataclasses.field(default=uint8, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class UInt16DType(UnsignedIntDType[uint16]):
+    scalar_type: Final[type[uint16]] = dataclasses.field(default=uint16, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class UInt32DType(UnsignedIntDType[uint32]):
+    scalar_type: Final[type[uint32]] = dataclasses.field(default=uint32, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class UInt64DType(UnsignedIntDType[uint64]):
+    scalar_type: Final[type[uint64]] = dataclasses.field(default=uint64, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class SignedIntDType(DType[SignedIntT]):
+    pass
+
+
+@dataclasses.dataclass(frozen=True)
+class Int8DType(SignedIntDType[int8]):
+    scalar_type: Final[type[int8]] = dataclasses.field(default=int8, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class Int16DType(SignedIntDType[int16]):
+    scalar_type: Final[type[int16]] = dataclasses.field(default=int16, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class Int32DType(SignedIntDType[int32]):
+    scalar_type: Final[type[int32]] = dataclasses.field(default=int32, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class Int64DType(SignedIntDType[int64]):
+    scalar_type: Final[type[int64]] = dataclasses.field(default=int64, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class FloatingDType(DType[FloatingT]):
+    kind: Final[Literal[DTypeKind.FLOAT]] = dataclasses.field(default=DTypeKind.FLOAT, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class Float32DType(FloatingDType[float32]):
+    scalar_type: Final[type[float32]] = dataclasses.field(default=float32, init=False)
+
+
+@dataclasses.dataclass(frozen=True)
+class Float64DType(FloatingDType[float64]):
+    scalar_type: Final[type[float64]] = dataclasses.field(default=float64, init=False)
+
+
+SliceLike = Union[int, tuple[int, ...], None, slice, "NDArrayObject"]
+
+NDArrayObject = Union[npt.NDArray, "NDArrayObjectProto"]
```

Comment:
I removed the additional cupy jax versions because I couldn't make mypy understand the conditional import pattern.